experiment/clustersecretbackup/main.go

```diff
@@ -0,0 +1,237 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	flag "github.com/spf13/pflag"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	_ "k8s.io/client-go/plugin/pkg/client/auth" // Enable all auth provider plugins
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/test-infra/experiment/clustersecretbackup/secretmanager"
+	"k8s.io/test-infra/gencred/pkg/util"
+
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	defaultSecretLabels = map[string]string{
+		"update_time": time.Now().Format("2006-01-02-15-04-05"),
+		"type":        "prow_backup",
+		"source":      "",
+	}
+)
+
+// options are the available command-line flags.
+type options struct {
+	project        string
+	clusterContext string
+	namespaces     []string
+	secrets        map[string]string
+	update         bool
+	dryRun         bool
+}
+
+type client struct {
+	kubeClient          ctrlruntimeclient.Client
+	secretmanagerClient secretmanager.ClientInterface
+	allSi               *corev1.SecretList
+	options
+}
+
+func (c *client) gsmSecretName(clusterSecret *corev1.Secret) string {
+	// Use cluster name, namespace and secret name is almost unique identifier.
+	// However, if consider GCP allow creating clusters with the same name under
+	// different zones, probably will need to add zones to this. Will address if
+	// ever needed.
+	return fmt.Sprintf("%s__%s__%s", c.clusterContext, clusterSecret.Namespace, clusterSecret.Name)
+}
+
+// gatherOptions parses the command-line flags.
+func gatherOptions(fs *flag.FlagSet, args ...string) options {
+	var o options
+	fs.StringVar(&o.project, "project", "", "GCP project used for backing up secrets")
+	fs.StringVar(&o.clusterContext, "cluster-context", "", "cluster context name used for backing up secrets, must be full form such as <PROVIDER>_<PROJECT>_<ZONE>_<CLUSTER>")
+	fs.StringSliceVar(&o.namespaces, "namespace", []string{}, "namespace to backup, can be passed in repeatedly")
+	fs.StringToStringVar(&o.secrets, "secret-name", nil, "namespace:name of secrets to be backed up, in the form of --secret-name=<namespace>=<name>")
```

Comment:
nit: Should mention that by default all secrets in the specified namespace(s) are backed up.

Reply:
done